### PR TITLE
5-GET/api/articles/article_id/comments

### PR DESCRIPTION
**Pull request by LukeRiches**
updated endpoints.js, added testing, endpoint, controller and model for 5-GET/api/articles/article_id/comments

### CODE REVIEW

File: __tests__/app.test.js [hunk 1]
```diff
@@ -87,6 +87,55 @@ describe('GET /api/articles/:article_id', () => {
...
 });
 
+describe('GET /api/articles/:article_id/comments', () => {
+  test('Should respond with an array of comments for the given article_id of which each comment should have the following properties: comment_id, votes, created_at, author, body, article_id', () => {
+    return request(app)
+      .get('/api/articles/1/comments')
+      .expect(200)
+      .then((response) => {
+        const comments = response.body;
+        console.log(comments, "comments");
```
Comment (two-triangles) [inline __tests__/app.test.js]: remove logs you don't need

File: __tests__/app.test.js [hunk 2]
```diff
@@ -87,6 +87,55 @@ describe('GET /api/articles/:article_id', () => {
...
+        expect(Array.isArray(comments)).toBe(true)
+
+        comments.forEach((comment) => {
```
Comment (two-triangles) [inline __tests__/app.test.js:157]: we should by also adding an assertion for checking the typeof the property values as well as checking that it has the right keys.

File: __tests__/app.test.js [hunk 3]
```diff
@@ -87,6 +87,55 @@ describe('GET /api/articles/:article_id', () => {
...
+  test('Sends an appropriate status and error message when given an invalid id', () => {
+    return request(app)
+      .get('/api/articles/999/comments')
+      .expect(404)
+      .then((response) => {
+        expect(response.body.msg).toBe('Article does not exist');
+      });
+  });
+});
+
```
Comment (two-triangles) [inline __tests__/app.test.js:208]: Test for invalid article_id is needed - 400 - bad request

File: db/Controllers/articles-controller.js [hunk 4]
```diff
@@ -13,4 +13,33 @@ function getArticleByID(req, res, next){
...
+    // })
+    // .catch((err)=>{
+    //     next(err);
+    // });
+
+    fetchArticleByID(article_id)
+    .then(()=>{
+        fetchCommentsByID(article_id)
+        .then((comments)=>{
+            console.log(comments);
```
Comment (two-triangles) [inline db/Controllers/articles-controller.js]: remove logs you don't need

File: db/Controllers/articles-controller.js [hunk 5]
```diff
@@ -13,4 +13,33 @@ function getArticleByID(req, res, next){
...
+            res.status(200).send(comments)
+        })
+        .catch((err)=>{
+            next(err);
```
Comment (two-triangles) [inline db/Controllers/articles-controller.js]: Your controller only needs one catch block. Your model returns a promise chain and if an error is thrown inside of it, then the catch block for your controller will catch it and pass it on to the correct error handling middleware

File: db/Models/articles-model.js [hunk 6]
```diff
@@ -15,5 +15,20 @@ function fetchArticleByID(article_id){
...
+function fetchCommentsByID(article_id){
+    let query = "SELECT * FROM comments WHERE article_id = $1 ORDER BY created_at;"
 
-module.exports = {fetchArticleByID};
\ No newline at end of file
+    return db.query(query, [article_id]).then(({rows}) => {
+        if (rows[0] === undefined) {
+            return Promise.reject({
+                status: 404, 
+                msg : "Article has no comments"
```
Comment (two-triangles) [inline db/Models/articles-model.js]: This error is too generic - 

 - 404 not found - article that doesn't exist 
 - 200 - empty array - article does exist but has no comments 
 
 This is the way that you should be testing this!

File: Controllers/articles-controller.js [hunk 7]
```diff
@@ -29,4 +29,31 @@ function getArticles(req, res, next){
...
+        .then((comments)=>{
+            res.status(200).send(comments)
+        })
+        .catch((err)=>{
+            next(err);
+        }); 
+    })
+    .catch((err)=>{
+        next(err);
+    });
```
Comment (Pipe-Smoking-Rabbit) [inline Controllers/articles-controller.js:56]: Hey Luke, you're really close with both methods ( _I can see you were trying Promise.all above_ ) 

If you have a look at the notes page on Promises [https://notes.northcoders.com/courses/js-fundamentals/promises](url) in the section titled **Benefits of Using Promises** there is an example of a promise chain using _.then_ blocks .

File: Models/articles-model.js [hunk 8]
```diff
@@ -31,7 +31,21 @@ function fetchArticles(){
     })
 }
 
+function fetchCommentsByID(article_id){
+    let query = "SELECT * FROM comments WHERE article_id = $1 ORDER BY created_at DESC;"
+
+    return db.query(query, [article_id]).then(({rows}) => {
+        if (rows[0] === undefined) {
+            return Promise.reject({
```
Comment (Pipe-Smoking-Rabbit) [inline Models/articles-model.js:39]: You shouldn't need this logic here, if there are no comments then returning `rows` will return `[]` which is all we need

**Review by Pipe-Smoking-Rabbit — APPROVED** [review]
Hey Luke, it's so so close, I've included a link to the notes where there is an example to look at. And another comment on your model, there's no need to include any logic for when there are no comments, just let it send back an empty array 👍

**Pull request merged**